core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -136,7 +137,24 @@ private void checkMappingsCompatibility(IndexMetaData indexMetaData) {
             // We cannot instantiate real analysis server at this point because the node might not have
             // been started yet. However, we don't really need real analyzers at this stage - so we can fake it
             IndexSettings indexSettings = new IndexSettings(indexMetaData, this.settings);
-            SimilarityService similarityService = new SimilarityService(indexSettings, null, Collections.emptyMap());
+            final Map<String, SimilarityProvider.Factory> similarityMap = new AbstractMap<String, SimilarityProvider.Factory>() {
+                @Override
+                public boolean containsKey(Object key) {
+                    return true;
+                }
+
+                @Override
+                public SimilarityProvider.Factory get(Object key) {
+                    assert key instanceof String : "key must be a string but was: " + key.getClass();
+                    return SimilarityService.BUILT_IN.get(SimilarityService.DEFAULT_SIMILARITY);
+                }
+
+                @Override
+                public Set<Entry<String, SimilarityProvider.Factory>> entrySet() {
+                    return Collections.emptySet();
```

Comment:
I realize the analyzer hack does this, but it is inconsistent with what is returned above. I would rather throw a UEO here so that if entrySet begins to be used, we will catch this rather than get weird behavior. (And we should, in a followup or in this PR, fix the other entrySet() impl as well).

Reply:
I can do that but need to know what an UEO is (I'm not a Java programmer).

Reply:
@xabbu42 I think that @rjernst meant `UOE` as in `UnsupportedOperationException`.

Reply:
Yes, Jason is correct on what I meant.

Reply:
I get test failures if make this change:

```
Suite: org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeServiceTests
  1> [2017-10-14T16:43:29,230][INFO ][o.e.c.m.MetaDataIndexUpgradeServiceTests] [testPluginUpgradeFailure]: before test
  1> [2017-10-14T16:43:29,233][INFO ][o.e.c.m.MetaDataIndexUpgradeServiceTests] [testPluginUpgradeFailure]: after test
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=67F6816A4DF09ADB -Dtests.class=org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeServiceTests -Dtests.method="testPluginUpgradeFailure" -Dtests.security.manager=true -Dtests.locale=sr -Dtests.timezone=Asia/Katmandu
FAILURE 0.03s | MetaDataIndexUpgradeServiceTests.testPluginUpgradeFailure <<< FAILURES!
   > Throwable #1: org.junit.ComparisonFailure: expected:<[unable to upgrade the mappings for the index [[foo/BOOM]]]> but was:<[Cannot upgrade index foo]>
   >    at __randomizedtesting.SeedInfo.seed([67F6816A4DF09ADB:37D52F4B7D45621E]:0)
   >    at org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeServiceTests.testPluginUpgradeFailure(MetaDataIndexUpgradeServiceTests.java:141)
   >    at java.lang.Thread.run(Thread.java:748)
  1> [2017-10-14T16:43:29,258][INFO ][o.e.c.m.MetaDataIndexUpgradeServiceTests] [testPluginUpgrade]: before test
  1> [2017-10-14T16:43:29,263][INFO ][o.e.c.m.MetaDataIndexUpgradeServiceTests] [testPluginUpgrade]: after test
```

Reply:
I think this happens because `IndexAnalyzers` iterates over the analyzers/normalizers when in `close()`. Go ahead and add back the empty map, but please add a comment noting the reason?